subsys/shell/shell_uart.c

```diff
@@ -111,6 +111,17 @@ static void uart_tx_handle(const struct device *dev,
 	len = ring_buf_get_claim(sh_uart->tx_ringbuf, (uint8_t **)&data,
 				 sh_uart->tx_ringbuf->size);
 	if (len) {
+		/* Wait for DTR signal before sending anything to output. */
```

Comment:
I believe we cannot do it each time. I think there shall be some `ifdef` to enable this code only for interfaces that are using flow control.

Reply:
No need for `ifdef`, my suggestion:

```diff
diff --git a/subsys/shell/Kconfig.backends b/subsys/shell/Kconfig.backends
index ceef58f361..5f4b4f6cc9 100644
--- a/subsys/shell/Kconfig.backends
+++ b/subsys/shell/Kconfig.backends
@@ -71,6 +71,13 @@ config SHELL_BACKEND_SERIAL_RX_POLL_PERIOD
        help
          Determines how often UART is polled for RX byte.
 
+config SHELL_BACKEND_SERIAL_CHECK_DTR
+       bool "Check DTR signal before TX"
+       default y if USB_UART_CONSOLE
+       depends on UART_LINE_CTRL
+       help
+         Check DTR signal before TX.
+
 module = SHELL_BACKEND_SERIAL
 default-timeout = 100
 source "subsys/shell/Kconfig.template.shell_log_queue_timeout"
diff --git a/subsys/shell/shell_uart.c b/subsys/shell/shell_uart.c
index 9945b283a2..f3d648fa50 100644
--- a/subsys/shell/shell_uart.c
+++ b/subsys/shell/shell_uart.c
@@ -101,6 +101,22 @@ static void uart_rx_handle(const struct device *dev,
        }
 }
 
+static void uart_check_dtr(const struct device *dev)
+{
+       if (IS_ENABLED(CONFIG_SHELL_BACKEND_SERIAL_CHECK_DTR)) {
+               int dtr;
+
+               while (true) {
+                       uart_line_ctrl_get(dev, UART_LINE_CTRL_DTR, &dtr);
+                       if (dtr) {
+                               break;
+                       } else {
+                               k_yield();
+                       }
+               };
+       }
+}
+
 static void uart_tx_handle(const struct device *dev,
                           const struct shell_uart *sh_uart)
 {
@@ -111,6 +127,7 @@ static void uart_tx_handle(const struct device *dev,
        len = ring_buf_get_claim(sh_uart->tx_ringbuf, (uint8_t **)&data,
                                 sh_uart->tx_ringbuf->size);
        if (len) {
+               uart_check_dtr(dev);
                len = uart_fifo_fill(dev, data, len);
                err = ring_buf_get_finish(sh_uart->tx_ringbuf, len);
                __ASSERT_NO_MSG(err == 0);
```

Reply:
@jfischer-no thank you for suggestion, I really like idea of putting it into Kconfig and creating separate method, so I introduced it. I didn't introduce k_yield, as it lead me to the state that application is blocked until the serial terminal will be opened, what doesn't seem to be correct. With k_sleep it seems to work smoothly.